pkg/util/deployment/deployment.go

```diff
@@ -444,3 +444,35 @@ func WaitForObservedDeployment(getDeploymentFunc func() (*extensions.Deployment,
 		return deployment.Status.ObservedGeneration >= desiredGeneration, nil
 	})
 }
+
+// ResolveFenceposts resolves both maxSurge and maxUnavailable. This needs to happen in one
+// step. For example:
+//
+// 2 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1), then old(-1), then new(+1)
+// 1 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1)
+// 2 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1)
+// 2 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1)
+func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired int) (int, int, error) {
+	surge, err := intstrutil.GetValueFromIntOrPercent(maxSurge, desired, true)
+	if err != nil {
+		return 0, 0, err
+	}
+	if surge == 0 {
+		// If a user has specified maxSurge to zero but maxUnavailable as a non-zero percent,
+		// then we want to allow scaling down and not block rollouts.
```

Comment:
Regarding round down max unavailable (I might be misunderstanding your point).

We don't want to go to zero pods in a max surge 0 case unless max
unavailability is 100%.  Otherwise rolling deployments will take
downtime and aren't flexible to scaling.  We want a naive user with 2
pods to only take downtime if they explicitly ask for 100%
unavailability.  If scale = 1 we can round down when max surge is 0
(since setting scale to 1 with surge 0 means you know you will take
downtime).